### PR TITLE
SFS-3757: Add frequency monitoring for USB_Cam Node

### PR DESCRIPTION
**Pull request by imjaya**
Added Frequency Monitoring for USB_Cam Node.

**Merge With**
https://github.com/MisoRobotics/miso_core/pull/393/
https://github.com/MisoRobotics/miso_peripherals/pull/154

### CODE REVIEW

**Review by hjoe72 — APPROVED** [review]

**Review by hungrh — CHANGES_REQUESTED** [review]

File: nodes/usb_cam_node.cpp [hunk 1]
```diff
@@ -98,13 +100,19 @@ class UsbCamNode
 
   UsbCamNode() :
       node_("~")
+      , min_frequency_{ framerate_*0.9 }
+      , max_frequency_{ framerate_*1.1 }
+      , frequency_status_{ diagnostic_updater::FrequencyStatusParam{ &min_frequency_, &max_frequency_ } }
```
[the remark this answers is not in the document]
Comment (imjaya) [reply nodes/usb_cam_node.cpp]: Thanks have addressed it in 436fc7c10de9470068efac168e955b2f77f93141

**Review by hungrh — APPROVED** [review]

**Review by rsinnet — DISMISSED** [review]
I am requesting changes because of a defect in a string. That being said. We could probably avoid diagnostics work on this node since it is third-party and won't track the upstream if we change it. Like, we're not going to go into MoveIt and add nodes there and we're not going to modify robot_state_publisher to publish diagnostics and so on, I should think, because we'd have to keep constantly pulling in upstream. IMO we should only add this to our code or third-party code that is known to crash.

I understand we have a new toy with this diagnostic logging, but we must be judicious about how we use it.

File: nodes/usb_cam_node.cpp [hunk 2]
```diff
@@ -343,11 +365,28 @@ class UsbCamNode
     ros::Rate loop_rate(this->framerate_);
     while (node_.ok())
     {
+      ros::Duration last_update_dur = ros::Time::now() - last_cam_update_time;
+      camera_status.name = "USB Camera Device Monitoring";
+      camera_status.hardware_id = "None";
+      if (last_update_dur > ros::Duration(loop_rate))
+    {
+      camera_status.level = diagnostic_msgs::DiagnosticStatus::ERROR;
+      camera_status.message = "Point Cloud not getting published";
```
[the remark this answers is not in the document]
Comment (imjaya) [reply nodes/usb_cam_node.cpp]: Ooop my bad, was copy pasting from my other PR. Fixed it in a36c5e140fc6340b92c84fcc024f93bfe566ce3b

File: nodes/usb_cam_node.cpp [hunk 3]
```diff
@@ -334,6 +355,7 @@ class UsbCamNode
 
     // publish the image
     image_pub_.publish(img_, *ci);
+    last_cam_update_time = ros::Time::now();
```
[the remark this answers is not in the document]
Comment (imjaya) [reply nodes/usb_cam_node.cpp]: Fixed.

**Comment by zzv2** [comment]
> > IMO we should only add this to our code or third-party code that is known to crash.
> > I understand we have a new toy with this diagnostic logging, but we must be judicious about how we use it.
> 
> I can understand your point of view @rsinnet, however by adding the Diagnostics here all we are doing is logging some information as an error identifier just in case if the node crashes and to see if the node is publishing at the desired rate. I cannot see why this might be difficult to manage.
> 
> @zzv2 your thoughts on this?

The camera nodes have been known to crash or malfunction occasionally so I think adding logging here makes sense. Also this is a small enough change that I don't think it would make merging in changes from upstream difficult.

File: nodes/usb_cam_node.cpp [hunk 4]
```diff
@@ -343,11 +366,28 @@ class UsbCamNode
...
+      if (last_update_dur > ros::Duration(loop_rate))
+    {
+      camera_status.level = diagnostic_msgs::DiagnosticStatus::ERROR;
+      camera_status.message = "USB Camera is not publishing Images.";
+    }
+    else
+    {
+      camera_status.level = diagnostic_msgs::DiagnosticStatus::OK;
+      camera_status.message = "Normal";
+    }
```
Comment (zzv2) [inline nodes/usb_cam_node.cpp:381]: i dont really see why this part is necessary. it seems like it duplicates the frequency status below

Comment (imjaya) [reply nodes/usb_cam_node.cpp]: That's true, I have removed it now. We thought we wanted a separate messaged key value pair for the Datadog Logger.

**Review by zzv2 — APPROVED** [review]

**Review by nardavin — APPROVED** [review]

**Review by zzv2 — APPROVED** [review]

**Pull request merged**